src/map/status_effect_container.cpp

```diff
@@ -178,16 +212,35 @@ uint8 CStatusEffectContainer::GetEffectsCount(EFFECT ID)
 {
     uint8 count = 0;
 
-    for (uint16 i = 0; i < m_StatusEffectList.size(); ++i)
+    for (CStatusEffect* PStatusEffect : m_StatusEffectSet)
     {
-        if (m_StatusEffectList.at(i)->GetStatusID() == ID && !m_StatusEffectList.at(i)->deleted)
+        if (PStatusEffect->GetStatusID() == ID && !PStatusEffect->deleted)
         {
             count++;
         }
     }
     return count;
 }
 
+uint8 CStatusEffectContainer::GetLowestFreeSlot()
+{
+    uint8 lowestFreeSlot = 1;
+
+    for (CStatusEffect* PStatusEffect : m_StatusEffectSet)
+    {
+        if (PStatusEffect->GetSlot() == lowestFreeSlot && !PStatusEffect->deleted)
+        {
+            lowestFreeSlot++;
+        }
+        else if (PStatusEffect->GetSlot() > lowestFreeSlot) {
+            // Can break since the set is sorted by slot number,
+            // and if we're past the lowest free one, we've found it already
+            break;
+        }
+    }
+    return lowestFreeSlot;
```

Comment:
Good suggestion. It could probably be changed to store the lowest free slot as a value in the container, but would then need some kind of data structure to make it easy to calculate the smallest free slot.

I didn't test all status effects, but at least songs and rolls are ordered like this. There could definitely be more like this that are grouped with songs and rolls, DNC buffs, that I haven't researched at this point in time. But this should be easy to add in later, if that's the case.

Reply:
After giving it some thought, I don't think it's worth the added complexity of changing this to keep track of lowest free slot, instead of just iterating whenever the lowest free slot is needed (which is not that often).

Besides, the code iterates through the status effects pretty much any time anything happens to some of them (i.e. see all the iterator changes I've made in the PR), so I don't think this is something that would be impacting performance in any noticeable way.